### PR TITLE
Issue #2833: backdrop_goto_deprecated() doesn't redirect wildcard paths.

### LINKED ISSUE #2833: Add support for wildcards in deprecated path redirects.
## Describe your issue or idea

We have used backdrop_goto_deprecated() in many places to handle old-to-new path changes. In most cases, these redirects work great. But if the redirect contains a wildcard, the redirect goes to a new URL with the name of the wildcard in the final URL itself, rather than performing the substitutions.

### Steps to reproduce (if reporting a bug)

- Install Backdrop.
- Go to a path that contains a wildcard redirect, such as `admin/structure/layouts/manage/%layout_tempstore/settings`. An example would be `admin/structure/layouts/manage/home/settings`.

### Actual behavior (if reporting a bug)

- User is redirected to `/admin/structure/layouts/manage/%25layout_tempstore/configure`

### Expected behavior (if reporting a bug)

- User should be redirected to `/admin/structure/layouts/manage/home/configure` (the wildcard `%layout_tempstore` replaced with `home`).

We have about a two dozen path redirects right now, and 6 of them seem to have this problem:

- `admin/structure/layouts/manage/%layout_tempstore/edit`
- `admin/structure/layouts/manage/%layout_tempstore/settings`
- `admin/config/search/redirect/edit/%redirect`
- `admin/config/search/redirect/delete/%redirect`
- `admin/config/search/path/delete/%path`
- `admin/config/search/path/edit/%path`

----

PR https://github.com/backdrop/backdrop/pull/1986

Advocate: @quicksketch

### PR DESCRIPTION
**Pull request by quicksketch**
Fixes https://github.com/backdrop/backdrop-issues/issues/2833.

### CODE REVIEW

File: core/modules/system/system.module [hunk 1]
```diff
@@ -3903,6 +3903,27 @@ function system_403() {
...
+ * This wrapper around backdrop_goto_deprecated() should be used as the menu
+ * callback in hook_menu() when redirecting from an old path to a new one.
+ *
+ * @since 1.8.0
+ */
+function system_redirect_deprecated_page($path) {
+  // Perform any substitutions in the passed-in path with the current path.
+  $parts = explode('/', $path);
+  foreach ($parts as $index => $path_part) {
+    if (substr($path_part, 0, 1) === '%') {
```
Comment (jlfranklin) [inline core/modules/system/system.module]: Should we use `substr()` or `backdrop_substr()`?  I don't think it matters, unless someone is running Backdrop with less common MB string encodings, like UTF16.  If the first character is "%", then it'll be the first byte in UTF8 as well as all (?) ISO 8859 code pages.  Either way, this and [#2800](/backdrop/backdrop-issues/issues/2800) should be consistent.

File: core/modules/system/system.module [hunk 2]
```diff
@@ -3903,6 +3903,27 @@ function system_403() {
...
+      $parts[$index] = arg($index);
+    }
```
Comment (jlfranklin) [inline core/modules/system/system.module:4304]: This assumes the parameters are in the same order and in the same place in both the old and new paths.  My first thought is to try to auto-map old to new if they're named (e.g., %node or %user), but that won't work with most paths.  What do you think about adding a second parameter to the function that maps the old to the new?  With that change, the `page arguments` would look something like this:

```php
  $menu['path/to/old/%/%/page/%'] = array(
    'page callback' => 'system_redirect_deprecated_page', 
    'page arguments' => array('path/%/to/%/new/page', array(1 => 4, 3 => 3)),
  );
```

If the second argument is missing, then a 1:1 mapping can be assumed.

**Review by jlfranklin — COMMENTED** [review]
This looks a lot better than the first go.

File: core/modules/system/system.module (hunk 1 above)
Comment (quicksketch) [reply core/modules/system/system.module]: Wouldn't hurt. I'll update.

File: core/modules/system/system.module (hunk 2 above)
Comment (quicksketch) [reply core/modules/system/system.module:4304]: Sure, that's not difficult to add and might be useful. It doesn't make it more difficult for existing 1:1 maps so seems like a general win.

File: core/modules/system/tests/system.test [hunk 3]
```diff
@@ -1007,6 +1007,63 @@ class PageNotFoundTestCase extends BackdropWebTestCase {
...
+   * Verifies that a watchdog message has been entered.
+   *
+   * @param $watchdog_message
+   *   The watchdog message.
+   * @param $variables
+   *   The array of variables passed to watchdog().
+   * @param $message
+   *   The assertion message.
+   */
+  private function assertWatchdogMessage($watchdog_message, $variables, $message) {
```
Comment (jlfranklin) [inline core/modules/system/tests/system.test]: Any reason this is here and not in the parent class `BackdropWebTestCase` with the other assert functions?  It seems way too useful for `DeprecatedRedirectPageCase` to keep for itself.

File: core/modules/path/path.module [hunk 4]
```diff
@@ -102,70 +102,70 @@ function path_menu() {
 
   // @deprecated: Remove backwards-compatibility redirect in 2.0.
   $items['admin/config/search/path'] = array(
-    'page callback' => 'backdrop_goto_deprecated',
+    'page callback' => 'system_redirect_deprecated_page',
```
Comment (jlfranklin) [inline core/modules/path/path.module:105]: So, the recommendation is to use `system_redirect_deprecated_page` even when there are no placeholders?  That's good.  The consistency will make it easy to apply as a rule in contrib module reviews.    We should include a brief note in the Backdrop Developer Documentation about the use of this function as a tool for providing backwards compatibility, and maybe an `@see` cross-reference in the `backdrop_goto_deprecated` docblock.

**Review by jlfranklin — APPROVED** [review]
Looks really good.  I have a couple things for your consideration, but otherwise I like it.

File: core/modules/system/tests/system.test (hunk 3 above)
Comment (quicksketch) [reply core/modules/system/tests/system.test]: Truth is I copied it directly from `CommentActionsTestCase`, so perhaps you're right that we should move it up into `BackdropWebTestCase` and save it from being private in two separate tests.

File: core/modules/path/path.module (hunk 4 above)
Comment (quicksketch) [reply core/modules/path/path.module:105]: Agreed. And thanks for reviewing this so quickly again!

**Review by jlfranklin — APPROVED** [review]
Looks great!

**Comment by quicksketch** [comment]
Could you mark the parent issue RTBC? Thanks so much @jlfranklin!

**Comment by jlfranklin** [comment]
I didn't see an RTBC.  Closest was "works for me."

**Comment by quicksketch** [comment]
Sorry, I should use the full label name: `pr - ready to be committed`

**Comment by jlfranklin** [comment]
> Sorry, I should use the full label name: pr - ready to be committed

LOL.  Different RTBC.  I was looking for "Reviewed and tested by the community."

**Comment by quicksketch** [comment]
Merged into 1.x for 1.19.0. Thanks @jlfranklin!

**Pull request merged**